File: Java/0003-longest-substring-without-repeating-characters.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int left = 0;
        int ret = 0;

        for (int index = 0; index < s.length(); index++) {
            char c = s.charAt(index);

            if (map.containsKey(c) && map.get(c) >= left) {
                left = map.get(c) + 1;
            }

            map.put(c, index);
            ret = Math.max(ret, index - left + 1);
        }

        return ret;
    }
}

File: Go/0003-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
	m := make(map[rune]int)
	left := 0
	ret := 0

	for index, char := range s {
		if prevIndex, exists := m[char]; exists && prevIndex >= left {
			left = prevIndex + 1
		}
		m[char] = index
		if currentLen := index - left + 1; currentLen > ret {
			ret = currentLen
		}
	}

	return ret
}